rfcs/solidity-snapps.md

```diff
@@ -0,0 +1,324 @@
+can be exposed at proof-generation time instead.
+
+### Struct and enum types
+
+[Struct types](https://docs.soliditylang.org/en/v0.8.5/structure-of-a-contract.html#struct-types)
+are a language feature of solidity, which is already supported by snarky.
+[Enum types](https://docs.soliditylang.org/en/v0.8.5/structure-of-a-contract.html#enum-types)
+have also had long-lived support in snarky, although have seen little use in
+practice.
+
+## Types
+
+### Value types
+
+All of the
+[value types](https://docs.soliditylang.org/en/v0.8.5/types.html#value-types)
+supported by solidity are also supported by snarky.
+
+### Reference types
+
+[Reference types](https://docs.soliditylang.org/en/v0.8.5/types.html#reference-types)
+in solidity refer to blocks of memory. These don't have a direct analog in
+snarks, but can be simulated -- albeit at a much higher computational cost --
+by cryptographic primitives. Many of these primitives are already implemented
+in snarky.
+
+### Mapping types
+
+[Mapping types](https://docs.soliditylang.org/en/v0.8.5/types.html#mapping-types)
+are similar to reference types, but associate some 'key' value with each data
+entry. This can be implemented naively on top of a primitive for
+`[(key, value)]`, an array of key-value pairs, which is already available in
+snarky.
+
+We currently do not support a primitive for de-duplication: a key may appear
+multiple times in a `[(key, value)]` implementation and the prover for a snapp
+could choose any of the values associated with the given key in a particular
+map. This will require some research into the efficiency of the different
+primitives available for this, but has no impact upon the transaction model.
+
+## Standard library
+
+### Block / transaction properties
+
+The
+[block and transaction properties](https://docs.soliditylang.org/en/v0.8.5/units-and-global-variables.html#block-and-transaction-properties)
+available in solidity are, at time of writing:
+* `blockhash(uint blockNumber)`
+  - Can be easily supported using the protocol state hash, although the timing
+    is tight for successful execution (receive a block, create the snapp proof,
+    send it, have it included in the next block).
+  - **Proposal:** support the previous `protocol_state_hash` in snapp
+    predicates.
+* `block.chainid`
+  - **Proposal:** expose the chain ID in the genesis constants, allow it to be
+    used in the snapp predicate.
+* `block.coinbase`
+  - Snapp proofs are generated before the block producer is known. Not possible
+    to support.
+* `block.difficulty`
+  - Not applicable, block difficulty is fixed.
+* `block.gaslimit`
+  - Not applicable, we don't have a gas model.
+* `block.number`
+  - Available by using `blockchain_length` in the snapp predicate.
+* `block.timestamp`
+  - Available by using `timestamp` in the snapp predicate.
+* `gasleft()`
+  - Not applicable, we don't have a gas model.
+* `msg.data`
+  - Available as part of the snapp input
+* `msg.sender`
```

Comment:
This is as easy as changing the "transaction commitment" used for the public input to include the fee_payer.

I will note that there is a bit of a trade-off here, as we'd lose the ability to have transactions that were proved independently and then sent off by someone. We could do both by including a bit for each party that says whether it wants to be verified against the public input that includes the fee-payer or the one that doesn't... but that adds some complexity

Reply:
Does it make sense to have it as a predicate then? This means its still accessible by querying the party if you want it, but otherwise gives separability.

Should we consider doing the equivalent for `parties` and the stack below the current party, so that a snapp can always be proved in isolation if it doesn't depend on these?